### PR TITLE
Highlight nodes containing searched text

### LINKED ISSUE #116: Highlight nodes containing searched text
Native platform supports fuzzy search feature that is used for filtering documents that contain nodes with particular text. Let's developer fuzzy search feature further and highlight found nodes if they are on viewport.

### PR DESCRIPTION
**Pull request by AdeThorMiwa**
Screenshot 2023-05-31 at 00 05 13

 closes #116

### CODE REVIEW

**Comment by AdeThorMiwa** [comment]
@StaffEngineer there's still a little issue with my implementation which i'm not exactly sure how to resolve, so if you check the PR, i'm basically cloning the result from the `fuzzy_search` func, however if i have multiple rect node with say an hello text, i'm guessing its suppose to highlight both of them if they are both within the viewport, but the `fuzzy_search` func only returns one of them, so it only highlight  a single one. not sure how to go about fixing this as i dont yet fully understand the `fuzzy_search` func yet

File: src/ui_plugin/systems/search.rs [hunk 1]
```diff
@@ -91,6 +95,12 @@ pub fn search_box_text_changed(
                 let result = fuzzy_search(index, str.as_str());
                 match result {
                     Ok(docs) => {
+                        let mut doc_ids = docs
+                            .clone()
+                            .into_iter()
+                            .map(|l| ReflectableUuid(l.node_id))
+                            .collect::<Vec<_>>();
```
Comment (StaffEngineer) [inline src/ui_plugin/systems/search.rs:102]: Let's use `std::collections::HashSet` to have O(1) amortised contains operation.

**Comment by StaffEngineer** [comment]
(answered about `fuzzy_search` func on discord)

The PR mostly looks great to me, nice job! The main thing to address before it's merged is handling when search text is empty. If search box is empty we have to unhighlight all previous highlighted nodes.

File: src/ui_plugin/systems/search.rs [hunk 2]
```diff
@@ -91,6 +95,12 @@ pub fn search_box_text_changed(
                 let result = fuzzy_search(index, str.as_str());
                 match result {
                     Ok(docs) => {
+                        let mut doc_ids = docs
```
Comment (StaffEngineer) [inline src/ui_plugin/systems/search.rs:98]: node_ids

Comment (StaffEngineer) [reply src/ui_plugin/systems/search.rs:98]: Not sure if it worth it but viewport can only have nodes from current document, so we could filter the nodes by that, current document can be obtained in AppState.

File: src/ui_plugin/systems/search.rs (hunk 1 above)
Comment (StaffEngineer) [reply src/ui_plugin/systems/search.rs:102]: Something like this:
```rust
let node_ids: HashSet<ReflectableUuid> = docs
    .into_iter()
    .filter(|l| l.doc_id == current_doc_id)
    .map(|l| ReflectableUuid(l.node_id))
    .collect()
```

**Comment by AdeThorMiwa** [comment]
> (answered about `fuzzy_search` func on discord)
> 
> The PR mostly looks great to me, nice job! The main thing to address before it's merged is handling when search text is empty. If search box is empty we have to unhighlight all previously highlighted nodes.

@StaffEngineer  The current implementation actually caters for that i believe, when you type in a value in the search box and some nodes were highlighted, if you empty the search box the highlighting reverts back

**Comment by StaffEngineer** [comment]
> Also, one of the check is failing, and i'm guessing this will require snapshot testing (or no) any pointers on how to fix that?

This is a check that coverage for the new code is higher than overall test coverage. For passing this check the new test should be added, see other tests in this file. I am fine with this check to be failing.

**Pull request merged**